### DISCUSSION TITLE
WiX4 Custom BA - is it possible?

### DISCUSSION THREAD

**Discussion by oleg-shilo** [wixtoolset/issues#7591] [Questions] [answered]
Is there any working example for Managed Custom BA for WiX 4.

The only information available for that is in WiX4 documentation. It states the following:

> Even though the WixBA is not currently shipping, it is still available at https://github.com/wixtoolset/wix4/tree/develop/src/test/burn/WixToolset.WixBA. You can basically see a step-by-step example of upgrading a v3 MBA by looking at the history of that folder and WixToolset.WixBA.csproj. Another folder that might be useful is https://github.com/wixtoolset/wix4/tree/develop/src/test/burn/TestData/TestBA. 

The link(s) above help with building the assembly with the BA but there is no guidance on how to embed this BA in the bundle.
This leaves developers with no working E2E code sample (WiX Bundle+BA project). My numerous attempts to solve it for the community have failed. 
Does anyone solve it?

I have created the test project: 
https://github.com/oleg-shilo/wixsharp/tree/wix-v4-master/Source/src/WixSharp.Samples/Wix%23%20Samples/Bootstrapper/WiX4-Spike/Bundle1

The sample can be built but fails to load at runtime:

```txt
[...][...]i000: Loading managed bootstrapper application.
[...][...]e000: Error 0x8013101b: Failed to create the bootstrapper application.
[...][...]e000: Error 0x8013101b: Failed to create the managed bootstrapper application.
[...][...]e000: Error 0x8013101b: Failed to create BA.
[...][...]e000: Error 0x8013101b: Failed to load BA.
[...][...]e000: Error 0x8013101b: Failed while running
```

**Comment by kerrywicks** [comment]
I don't have an example I can share, but I've converted a Wix3 BA to Wix 4. I had a quick glance at your repo, and I *think* you need to set up an AssemblyInfo.cs for your WixBA project, with the following: 

`
[assembly: BootstrapperApplicationFactory(typeof(WixBAFactory))]
`

**Reply by oleg-shilo** [reply]
Unfortunately, it is not the cause.

The attribute is set. The assembly I am using ([WixToolset.WixBA](https://github.com/wixtoolset/wix4/tree/develop/src/test/burn/WixToolset.WixBA)) is the BA recommended by the WiX team and it does have its attribute set. I use it without any change.

**Reply by kerrywicks** [reply]
I had some time to have a poke at this to get it working. I think it's your config file (which isn't included in your source code when I looked at your branch). 

Create WixBA.BootstrapperCore.config under WixToolset.WixBA. It's contents should be the same as what you can find in the example code here: https://github.com/wixtoolset/Bal.wixext/blob/master/src/test/examples/FullFramework4MBA/WixToolset.Mba.Host.config

If you did a copy/paste from the file above, replace this line:

` <host assemblyName="Example.FullFramework4MBA" />`

with 

` <host assemblyName="WixToolset.WixBA" />`

That should get you up and running.

**Comment by oleg-shilo** [comment]
Unfortunately, it's not it.

I, in fact, used that `WixBA.BootstrapperCore.config` locally. I only accidentally missed it when committing the repo.
I added it now (https://github.com/oleg-shilo/wixsharp/blob/wix-v4-master/Source/src/WixSharp.Samples/Wix%23%20Samples/Bootstrapper/WiX4-Spike/Bundle1/WixToolset.WixBA/WixBA.BootstrapperCore.config)

Did you say "a poke at this to get it working"? If it is the case can you just share the complete working project?
I feel that is is something very simple and I just cannot see it.

**Reply by kerrywicks** [reply]
I no longer have the working project - I literally took 10 minutes of my spare time, added the config file, built the bootstrapper in Debug x86, the package in debug x64 (as that is the output you were using in the bundle), and the bundle in debug x86, and it worked. I have since deleted it, as it was taking up more space than I liked and the fix for your issue seemed pretty simple. I made no other changes. Your config seems correct now.

If you're using visual studio, I'd advise doing a full solution 'clean', as sometimes a build/rebuild fails to clean out older builds with the Wix stuff, I've found. If you're doing a rebuild, I'd advise trying to get everything built on the same configuration.

If you're still struggling after that, I'd recommend having a look at the example projects here: https://github.com/wixtoolset/Bal.wixext/tree/master/src/test/examples as they seem to be a bit simpler than what you're working with at the moment.

**Comment by oleg-shilo** [comment]
Bingo!
Thank you for the hint. The problem was in the _Platform_ setting but the code itself was OK all along.

This is the config that I was trying to use (did not work):
![image](https://github.com/wixtoolset/issues/assets/16729806/30f720b6-4474-4606-8525-1d746646be41)

And this change made it work:

![image](https://github.com/wixtoolset/issues/assets/16729806/da347e08-429c-41db-814d-d3024ccc5f22)

Thank you for your help. Really appreciate it.

**Reply by kerrywicks** [reply, marked as answer]
I'm glad you got it working in the end - good luck with the rest!

**Comment by oleg-shilo** [comment]
Adding the project to assist others:
[Bundle1.zip](https://github.com/wixtoolset/issues/files/11983570/Bundle1.zip)

**Reply by Shivapatne** [reply]
do you have this example for Wix 5?

**Reply by oleg-shilo** [reply]
No, I don't.

Wix5 brings very positive but breaking changes around custom BA. The problem is that at the time of Wix5 release it did not have any example of a custom BA app. So WixSharp did not migrate to new BA architecture and falls back to Wix4 compiler when it detects that user is trying to build a custom BA

I am not sure if WiX team has now provided the guidance/example of custom BA for Wix5. If they did then it can definitely be integrated into WiSharp.

**Reply by oleg-shilo** [reply]
Good news. there is a sample: https://wixtoolset.org/docs/fivefour/oopbas/

Note for myself :)
This sample is still for the WiX4 in-process BA, not the new one: https://github.com/wixtoolset/wix/tree/HEAD/src/test/burn/WixToolset.WixBA### 

I will schedule this as a new feature for the next release.

**Reply by Shivapatne** [reply]
> Adding the project to assist others: [Bundle1.zip](https://github.com/wixtoolset/issues/files/11983570/Bundle1.zip)

Is this project upgrading successfully but why is it requiring us to click the install button again for the upgrade?  why it is not showing the upgrade button ?

**Reply by oleg-shilo** [reply]
The sample I shared is for wix4 and just a sample that I isolated from this repo codebase. 
The sample may or may not work fully meet the user expectations. I am not the author of the solution, so cannot judge. 
I was mostly interested in finding a way of building other custom BA's based on this sample. So for me that sample was a huge help so I shared it here.

As for now, I believe WiX team has already released the sample for WiX5 so we all should switch to it instead.

**Reply by Shivapatne** [reply]
Thank you so much! Your sample code has been a great help. Could you let me know where the WiX team or you have provided the source code sample for WiX 5?

**Reply by oleg-shilo** [reply]
You are very welcome.

> Could you let me know where the WiX team...

You are already here. :)
This repository is THE workplace of the WiX team. Though I am just a guest here the same as you are.

> you have provided the source code sample for WiX 5

I have captured the WiX5 sample from this repo some time ago and included in my repo [here](https://github.com/oleg-shilo/wixsharp/tree/master/Source/src/WixSharp.Samples/Wix%23%20Samples/Bootstrapper/WiX5-Spike/WixToolset.WixBA).

But most likely WiX team has developed it further and you can find the latest version in this very repository.

**Reply by Shivapatne** [reply]
“Do you have any example projects that use WiX 6.0 with a Bootstrapper Application (BA) implemented in .NET Core? I’m exploring this setup and would appreciate any reference or sample.”

this is my code :
		<BootstrapperApplication SourceFile="..\WixToolset.WixBA\bin\Debug\net6.0-windows\win-x86\WixToolset.WixBA.exe">
			<Payload SourceFile="..\WixToolset.WixBA\bin\Debug\net6.0-windows\win-x86\WixToolset.WixBA.deps.json" />
			<Payload SourceFile="..\WixToolset.WixBA\bin\Debug\net6.0-windows\win-x86\WixToolset.WixBA.dll" />
			<Payload SourceFile="..\WixToolset.WixBA\bin\Debug\net6.0-windows\win-x86\WixToolset.WixBA.runtimeconfig.json" />
			<Payload SourceFile="..\WixToolset.WixBA\bin\Debug\net6.0-windows\win-x86\mbanative.dll" />
			<Payload SourceFile="..\WixToolset.WixBA\bin\Debug\net6.0-windows\win-x86\WixToolset.BootstrapperApplicationApi.dll"/>
			<Payload SourceFile="..\WixToolset.WixBA\bin\Debug\net6.0-windows\win-x86\WixToolset.Mba.Core.dll" />
		</BootstrapperApplication>
still its not hitting to the BA 

<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs" xmlns:bal="http://wixtoolset.org/schemas/v4/wxs/bal">
	<Bundle Name="Bundle1" Manufacturer="TODO Manufacturer" Version="1.0.0.1" UpgradeCode="1c4f0354-a1f2-4f01-99d1-eec5f77ab98b">
		

		<BootstrapperApplication SourceFile="..\WixToolset.WixBA\bin\Debug\net6.0\WixToolset.WixBA.exe">
			<Payload SourceFile="..\WixToolset.WixBA\bin\Debug\net6.0\WixToolset.WixBA.deps.json" />
			<Payload SourceFile="..\WixToolset.WixBA\bin\Debug\net6.0\WixToolset.WixBA.dll" />
			<Payload SourceFile="..\WixToolset.WixBA\bin\Debug\net6.0\WixToolset.WixBA.runtimeconfig.json" />
			<Payload SourceFile="..\WixToolset.WixBA\bin\Debug\net6.0\mbanative.dll" />
			<Payload SourceFile="..\WixToolset.WixBA\bin\Debug\net6.0\WixToolset.Mba.Core.dll" />
		</BootstrapperApplication>

		

		<Chain>
			
		
			<MsiPackage bal:PrereqPackage="yes" SourceFile="..\Package1\bin\x64\Debug\en-US\Package1.msi" />
		</Chain>
	</Bundle>
</Wix>

**Reply by oleg-shilo** [reply]
The complete example of the external BA (including .NET Core) can be found in the samples folder of this repository:
<img width="383" height="185" alt="image" src="https://github.com/user-attachments/assets/a35138ee-cc53-4179-b54c-774280befd51" />

But I strongly advise you against targeting . NET Core family as your BA may not work on the target system since you have no guarantee that .NET Core is installed there.

**Reply by Shivapatne** [reply]
Yes, I agree with your observation. Could you please suggest how to overcome this issue or share any bundle examples for reference?
<img width="245" height="77" alt="image" src="https://github.com/user-attachments/assets/af6ef2bf-e3ab-4702-b2b1-d9e7e3e66665" />

**Reply by oleg-shilo** [reply]
You cannot. It's a logical constraint. 
A setup application is implemented to bring external applications/modules on the host OS. Thus, a setup application cannot have any dependency on external modules. It must be possible to run the setup app in a raw OS. 

This is the reason BA must be either native or .NET Framework (part of OS)

**Reply by robmen** [reply]
FYI: @oleg-shilo is incorrect. You can provide "primary" and "secondary" BootstrapperApplications. If your BAs prereqs are always present, a single BootstrapperApplication is enough. If not, you can use the `WixPrerequisiteBootstrapperApplication` (or your own BA without prereqs) to install prereqs. Then you can choose whether to make the "prereq BA" primary (as in it always tries to install the dependencies first) and your custom BA "secondary" (so it always runs after the "prereq BA") _or__ you can always try to run your BA as "primary" and if it fails, fallback to the "secondary BA" to install prereqs. Burn has logic to toggle between them P -> S -> P for exactly this scenario.

Like all things bundle-related, end-to-end documentation is sorely lacking. So the examples in the WiX test often show best. [This example has the custom BA as primary](https://github.com/wixtoolset/wix/tree/main/src/test/burn/TestData/PrereqBaTests/BundleA) and [this one has the custom BA as secondary](https://github.com/wixtoolset/wix/tree/main/src/test/burn/TestData/PrereqBaTests/BundleC).

Note: wixtoolset/issues#8654 is open to add more documentation around `Primary` and `Secondary` as part of my WiX v7 work. I was just looking at this issue this weekend while fixing a couple of other issues. 😄

**Reply by oleg-shilo** [reply]
@Shivapatne, what Rob is suggesting is quite possibly the workaround you are looking for. I suggest you consider it if the benefits justify dealing with two separate BAs .

@robmen, I was referring to the single BA scenario, which is indeed a subject of the constraint -  all prerequisites need to be present on the target system. 
This is the case the very case that is described as "your own BA without prereqs". 

@Shivapatne, I posted my response to you unintentionally. I received a GitHub notification and missed the point that it was not coming from the WixSharp (my project) but from the WiX Toolset.
Thus, I should not have posted the WixSharp example here, as it may only confuse you. I apologise for that.

**Reply by Shivapatne** [reply]
@oleg-shilo Have you tried this workaround? I tried it but wasn’t successful — please share the code if you’ve tried it.

**Reply by oleg-shilo** [reply]
No, I did not. I only commented on the suggestion as a valid one.
However, if I were facing the same challenge, I would rather avoid the dependency of the BA on .NET Core and stick with a single BA approach. So I would not need to use that workaround.

**Reply by Shivapatne** [reply]
@robmen Is your implementation applied after clicking the Install button in the UI? Please correct me if I’m wrong. However, the issue is that the UI itself is not launching because the machine does not have .NET Core installed.

**Reply by robmen** [reply]
No. The secondary BA automatically launches after the primary BA fails to start. The secondary BA can install pre-reqs, like .NET Core, then return control to the primary BA. Try the examples I mentioned above. We did the hard work. 😄

**Reply by Shivapatne** [reply]
@robmen Thanks a lot. Yes, it’s working fine, but I want to install it without the UI of the secondary BA.

<img width="505" height="347" alt="image" src="https://github.com/user-attachments/assets/0925d208-8fe3-4f7a-9fba-6777d9665ffc" />
